stew/results.nim

```diff
@@ -542,111 +770,84 @@ func error*[T, E](self: Result[T, E]): E =
       raiseResultDefect("Trying to access error when value is set", self.v)
     else:
       raiseResultDefect("Trying to access error when value is set")
-  self.e
-
-template value*[T, E](self: Result[T, E]): T =
-  mixin get
-  self.get()
-
-template value*[T, E](self: var Result[T, E]): T =
-  mixin get
-  self.get()
-
-template valueOr*[T, E](self: Result[T, E], def: T): T =
+  when E isnot void:
+    self.e
+
+func tryError*[T, E](self: Result[T, E]): E {.inline.} =
+  ## Fetch error of result if set, or raise
+  ## Raises a ResultError[T]
+  mixin raiseResultOk
+  if self.o: self.raiseResultOk()
+  when E isnot void:
+    self.e
+
+# Alternative spellings for get
+template value*[T, E](self: Result[T, E]): T = self.get()
+template value*[T: not void, E](self: var Result[T, E]): var T = self.get()
+
+template valueOr*[T: not void, E](self: Result[T, E], def: T): T =
   ## Fetch value of result if set, or supplied default
   ## default will not be evaluated iff value is set
   if self.o: self.v
   else: def
 
-# void support
-
-template ok*[E](R: type Result[void, E]): auto =
-  ## Initialize a result with a success and value
-  ## Example: `Result[int, string].ok(42)`
-  R(o: true)
-
-template ok*[E](self: var Result[void, E]) =
-  ## Set the result to success and update value
-  ## Example: `result.ok(42)`
-  mixin ok
-  self = (type self).ok()
-
-template ok*(): auto =
-  mixin ok
-  ok(typeof(result))
-
-template err*(): auto =
-  mixin err
-  err(typeof(result))
-
-# TODO:
-# Supporting `map` and `get` operations on a `void` result is quite
-# an unusual API. We should provide some motivating examples.
-
-func map*[E, A](
-    self: Result[void, E], f: proc(): A): Result[A, E] {.inline.} =
-  ## Transform value using f, or return error
-  if self.o: result.ok(f())
-  else: result.err(self.e)
-
-func flatMap*[E, A](
-    self: Result[void, E], f: proc(): Result[A, E]): Result[A, E] {.inline.} =
-  if self.o: f(self.v)
-  else: Result[A, E].err(self.e)
+func flatten*[T, E](self: Result[Result[T, E], E]): Result[T, E] =
+  ## Remove one level of nesting
+  if self.o:
+    self.v
+  else:
+    when E is void:
+      err(Result[T, E])
+    else:
+      err(Result[T, E], self.error)
 
-func mapErr*[E, A](
-    self: Result[void, E], f: proc(x: E): A): Result[void, A] {.inline.} =
-  ## Transform error using f, or return value
-  if self.o: result.ok()
-  else: result.err(f(self.e))
+func filter*[T, E](
```

Comment:
I personally wouldn't want to see this monadic style handling errors being overused in the project.

Reply:
the core motivation here is to allow `Opt` to offer the same API as `Option` - the function definition is a generalization of the same concept